In this assignment, you will practice loading data into your python environment. It is recommended that you use either google colab or juypter notebook, however, you may use any environment you prefer as long as you are using the pandas library.

From the following link, choose two files to answer the questions below. https://github.com/data602sps/datasetspractice

In addition, you will also search the internet to find two datasets to answer the questions below.

Websites to find data:
- https://opendata.cityofnewyork.us/
- https://datasetsearch.research.google.com/
- https://www.kaggle.com/datasets
- http://archive.ics.uci.edu/ml/index.php
- https://data.gov/

**Q1**. Load your dataset into python using the pandas library and save data into a dataframe named dfi (where i is one of your datasets, for a total of 4).

In [165]:
import pandas as pd

df1 = pd.read_csv('https://raw.githubusercontent.com/data602sps/datasetspractice/main/kobe.csv')
df2 = pd.read_csv('https://raw.githubusercontent.com/data602sps/datasetspractice/main/salarydata.csv')
df3 = pd.read_csv('https://raw.githubusercontent.com/ShanaFarber/cuny-sps/master/DATA_602/assignment6/disney_plus_shows.csv')
df4 = pd.read_csv('https://raw.githubusercontent.com/ShanaFarber/cuny-sps/master/DATA_602/assignment6/products_data.csv')

**Q2**. Preview your data by calling your dataframe’s name. How many columns and rows do you see?

In [166]:
df1

vs  game quarter     time  \
0    ORL     1       1     9:47   
1    ORL     1       1     9:07   
2    ORL     1       1     8:11   
3    ORL     1       1     7:41   
4    ORL     1       1     7:03   
..   ...   ...     ...      ...   
128  ORL     3       4     3:33   
129  ORL     3       4     2:02   
130  ORL     3       4  00:23.9   
131  ORL     3       4  00:06.9   
132  ORL     3       4  00:00.5   

                                           description basket  
0              Kobe Bryant makes 4-foot two point shot      H  
1                            Kobe Bryant misses jumper      M  
2                     Kobe Bryant misses 7-foot jumper      M  
3    Kobe Bryant makes 16-foot jumper (Derek Fisher...      H  
4                      Kobe Bryant makes driving layup      H  
..                                                 ...    ...  
128                        Bryant Layup Shot: Missed        M  
129                          Bryant 3pt Shot: Missed        M  
130                          Bryant 3pt Shot: Missed        M  
131                          Bryant 3pt Shot: Missed        M  
132                 Bryant Layup Shot: Made (31 PTS)        H  

[133 rows x 6 columns]

In [167]:
df2

Sate  2018 Mean Annual Wage  Unnamed: 2
0               Alabama                 90,620         NaN
1                 Alaska                99,180         NaN
2                Arizona                88,800         NaN
3               Arkansas                81,430         NaN
4             California                97,110         NaN
5               Colorado                82,560         NaN
6            Connecticut                96,010         NaN
7               Delaware                93,880         NaN
8   District of Columbia                89,880         NaN
9                Florida                87,410         NaN
10               Georgia                86,320         NaN
11                Hawaii                90,540         NaN
12                 Idaho                77,700         NaN
13              IIIinois                90,690         NaN
14               lndiana                83,680         NaN
15                  lowa                82,960         NaN
16                Kansas                85,250         NaN
17              Kentucky                84,630         NaN
18             Louisiana                89,860         NaN
19                 Maine                79,910         NaN
20              Maryland                85,170         NaN
21         Massachusetts                91,750         NaN
22              Michigan                91,160         NaN
23             Minnesota                83,750         NaN
24           Mississippi                89,720         NaN
25              Missouri                81,330         NaN
26               Montana                79,050         NaN
27              Nedraska                80,130         NaN
28                Nevada               107,920         NaN
29         New Hampshire                82,880         NaN
30            New Jersey                97,770         NaN
31           New  Mexico                97,210         NaN
32              New York                87,470         NaN
33        North Carolina                87,560         NaN
34          North Dakota                78,120         NaN
35                  Ohio                86,690         NaN
36              Okahoma                 84,860         NaN
37                Oregon                85,890         NaN
38          Pennsylvania                87,050         NaN
39          Rhode Island                83,850         NaN
40        South Carolina                85,450         NaN
41          South Dakota                92,940         NaN
42             Tennessee                85,940         NaN
43                 Taxas                92,940         NaN
44                  Utah                85,940         NaN
45               Vermont                75,010         NaN
46              Virginia                91,700         NaN
47            Washington                85,930         NaN
48         West Virginia                89,420         NaN
49             Wisconsin                85,200         NaN
50               Wyomimg                87,510         NaN

In [168]:
df3

imdb_id                                       title  \
0    tt0147800                  10 Things I Hate About You   
1    tt7019028                        101 Dalmatian Street   
2    tt0115433                              101 Dalmatians   
3    tt0324941  101 Dalmatians 2: Patch's London Adventure   
4    tt0211181                              102 Dalmatians   
..         ...                                         ...   
987  tt3544734                                      Zapped   
988  tt0186726             Zenon: Girl of the 21st Century   
989  tt0271271                           Zenon: The Zequel   
990  tt0379060                                   Zenon: Z3   
991  tt2948356                                    Zootopia   

                                                  plot    type  rated   year  \
0    A pretty, popular teenager can't go out on a d...   movie  PG-13   1999   
1    This series follows the lives of Delilah and D...  series    NaN  2018–   
2    An evil high-fashion designer plots to steal D...   movie      G   1996   
3    Being one of 101 takes its toll on Patch, who ...   movie      G   2002   
4    Cruella DeVil gets out of prison and goes afte...   movie      G   2000   
..                                                 ...     ...    ...    ...   
987  When Zoey's mom remarries, Zoey finds it hard ...   movie   TV-G   2014   
988  An inquisitive teenager pries into suspected d...   movie   TV-G   1999   
989  A mischievous fifteen-year-old girl living on ...   movie   TV-G   2001   
990  The heroine sets out to win an important compe...   movie   TV-G   2004   
991  In a city of anthropomorphic animals, a rookie...   movie     PG   2016   

     released_at           added_at  runtime  \
0    31 Mar 1999  November 12, 2019   97 min   
1    25 Mar 2019  February 28, 2020      NaN   
2    27 Nov 1996  November 12, 2019  103 min   
3    21 Jan 2003  November 12, 2019   74 min   
4    22 Nov 2000  November 12, 2019  100 min   
..           ...                ...      ...   
987  23 Jun 2014  November 12, 2019  102 min   
988  23 Jan 1999  November 12, 2019   97 min   
989  12 Jan 2001  November 12, 2019  100 min   
990  11 Jun 2004  November 12, 2019   81 min   
991  04 Mar 2016  November 12, 2019  108 min   

                                                 genre  \
0                               Comedy, Drama, Romance   
1                            Animation, Comedy, Family   
2                     Adventure, Comedy, Crime, Family   
3        Animation, Adventure, Comedy, Family, Musical   
4                            Adventure, Comedy, Family   
..                                                 ...   
987                            Comedy, Family, Fantasy   
988                  Adventure, Comedy, Family, Sci-Fi   
989                  Comedy, Family, Adventure, Sci-Fi   
990                  Adventure, Comedy, Family, Sci-Fi   
991  Animation, Adventure, Comedy, Crime, Family, M...   

                                              director  \
0                                           Gil Junger   
1                                                  NaN   
2                                        Stephen Herek   
3                            Jim Kammerud, Brian Smith   
4                                           Kevin Lima   
..                                                 ...   
987                                      Peter DeLuise   
988                                    Kenneth Johnson   
989                                         Manny Coto   
990                                         Steve Rash   
991  Byron Howard, Rich Moore, Jared Bush(co-director)   

                                                writer  \
0                        Karen McCullah, Kirsten Smith   
1                                                  NaN   
2        Dodie Smith (novel), John Hughes (screenplay)   
3    Jim Kammerud (story), Dan Root (story), Garret...   
4    Dodie Smith (novel), Kristen Buckle

In [169]:
df4

Name Current Price Previous Price  \
0                          English Cucumber          1.99           1.99   
1                               Green Onion          1.29           1.29   
2           Bi-Colour Corn, Corn on the Cob          0.99           0.99   
3                     Romaine Heart, 3 Pack          6.00           6.00   
4                           Red Peppers              2.38           2.38   
...                                     ...           ...            ...   
17254                    Beef Sausage Rolls          8.79           8.79   
17255  Creamy Chicken and Vegetable Pot Pie          8.00           8.00   
17256                     Quiche Florentine          8.00           8.00   
17257                             Green Tea          2.29           2.29   
17258             Hard-boiled Eggs, Omega-3          1.29           1.29   

       Price Per Each          Category      Product ID  \
0           $1.99/1ea  Fresh Vegetables  20070132001_EA   
1           $1.29/1ea  Fresh Vegetables  20107500001_EA   
2           $0.99/1ea  Fresh Vegetables  20131170001_EA   
3           $2.00/1ea  Fresh Vegetables  20067389001_EA   
4           $9.90/1kg  Fresh Vegetables  20007535001_KG   
...               ...               ...             ...   
17254      $1.40/100g    Quiches & Pies     21116219_EA   
17255  $1,600.00/100g    Quiches & Pies     21401776_EA   
17256      $1.60/100g    Quiches & Pies     21401761_EA   
17257     $0.48/100ml         Beverages     20923222_EA   
17258      $1.47/100g     Snacks & Dips     21049546_EA   

                                             Product URL Amount  Units  
0      https://www.realcanadiansuperstore.ca/english-...      1     ea  
1      https://www.realcanadiansuperstore.ca/green-on...      1  bunch  
2      https://www.realcanadiansuperstore.ca/bi-colou...      1     ea  
3      https://www.realcanadiansuperstore.ca/romaine-...      3   pack  
4      https://www.realcanadiansuperstore.ca/red-pepp...      1     ea  
...                                                  ...    ...    ...  
17254  https://www.realcanadiansuperstore.ca/beef-sau...    630      g  
17255  https://www.realcanadiansuperstore.ca/creamy-c...    500      g  
17256  https://www.realcanadiansuperstore.ca/quiche-f...    500      g  
17257  https://www.realcanadiansuperstore.ca/green-te...    473     ml  
17258  https://www.realcanadiansuperstore.ca/hard-boi...     88      g  

[17259 rows x 9 columns]

`df1` has 133 rows and 6 columns. `df2` has 51 rows and 3 columns. `df3` has nd 19 columns. `df4` has 17,259 rows and 9 columns. 

**Q3**. Examine the shape of your data using the `.shape` command and the data types of your columns using `.dtypes`.

In [170]:
df1.shape

(133, 6)

In [171]:
df1.dtypes

vs             object
game            int64
quarter        object
time           object
description    object
basket         object
dtype: object

In [172]:
df2.shape

(51, 3)

In [173]:
df2.dtypes

Sate                      object
2018 Mean Annual Wage     object
Unnamed: 2               float64
dtype: object

In [174]:
df3.shape

(992, 19)

In [175]:
df3.dtypes

imdb_id         object
title           object
plot            object
type            object
rated           object
year            object
released_at     object
added_at        object
runtime         object
genre           object
director        object
writer          object
actors          object
language        object
country         object
awards          object
metascore      float64
imdb_rating    float64
imdb_votes      object
dtype: object

In [176]:
df4.shape

(17259, 9)

In [177]:
df4.dtypes

Name              object
Current Price     object
Previous Price    object
Price Per Each    object
Category          object
Product ID        object
Product URL       object
Amount            object
Units             object
dtype: object

**Q4**. Use `.describe()` on your data. What do you notice about your data? What does this command return?

In [178]:
df1.describe()

game
count  133.000000
mean     2.902256
std      1.450521
min      1.000000
25%      1.000000
50%      3.000000
75%      4.000000
max      5.000000

In [179]:
df2.describe()

Unnamed: 2
count         0.0
mean          NaN
std           NaN
min           NaN
25%           NaN
50%           NaN
75%           NaN
max           NaN

In [180]:
df3.describe()

metascore  imdb_rating
count  292.000000   879.000000
mean    62.061644     6.656428
std     15.776455     1.020352
min     19.000000     1.500000
25%     51.000000     6.100000
50%     61.000000     6.700000
75%     73.000000     7.400000
max     99.000000     9.700000

In [181]:
df4.describe()

Name Current Price Previous Price Price Per Each        Category  \
count       17259         17259          17259          17259           17259   
unique      13147           817            829           1527             118   
top     Soy Sauce          4.99           4.99     $1.50/100g  Chips & Snacks   
freq           17           823            823            189             951   

         Product ID                                        Product URL Amount  \
count         17259                                              17259  17259   
unique        14423                                              14423    982   
top     21190251_EA  https://www.realcanadiansuperstore.ca/apples-c...      1   
freq             14                                                 14   1513   

        Units  
count   17259  
unique     12  
top         g  
freq    10792

`.describe()` returns the descriptive statistics for columns in the dataframe. 

**Q5**. Use the `.head()` and `.tail()` command - what does this do?

In [182]:
df1.head()

vs  game quarter  time                                        description  \
0  ORL     1       1  9:47            Kobe Bryant makes 4-foot two point shot   
1  ORL     1       1  9:07                          Kobe Bryant misses jumper   
2  ORL     1       1  8:11                   Kobe Bryant misses 7-foot jumper   
3  ORL     1       1  7:41  Kobe Bryant makes 16-foot jumper (Derek Fisher...   
4  ORL     1       1  7:03                    Kobe Bryant makes driving layup   

  basket  
0      H  
1      M  
2      M  
3      H  
4      H

In [183]:
df1.tail()

vs  game quarter     time                         description basket
128  ORL     3       4     3:33         Bryant Layup Shot: Missed        M
129  ORL     3       4     2:02           Bryant 3pt Shot: Missed        M
130  ORL     3       4  00:23.9           Bryant 3pt Shot: Missed        M
131  ORL     3       4  00:06.9           Bryant 3pt Shot: Missed        M
132  ORL     3       4  00:00.5  Bryant Layup Shot: Made (31 PTS)        H

`.head()` shows the top n rows of the dataframe. `.tail()` shows the bottom n rows of the dataframe.  'n' defaults to 5 unless a different number is given in the parenthesis.

**Extra Credit (3 pts)**

1. Choose one of your datasets and remove the header information. (Can delete the row in excel, etc..)
2. Import the data into your environment using pandas. Display the .head() of your data showing no header information.
3. Using pandas, update the dataset to include the header information. Display the updated data using .head().

In [184]:
# import without header
df4_ec = pd.read_csv('https://raw.githubusercontent.com/ShanaFarber/cuny-sps/master/DATA_602/assignment6/products_data-no_header.csv', header=None)

df4_ec.head()

0      1      2          3                 4  \
0                 English Cucumber   1.99   1.99  $1.99/1ea  Fresh Vegetables   
1                      Green Onion   1.29   1.29  $1.29/1ea  Fresh Vegetables   
2  Bi-Colour Corn, Corn on the Cob   0.99   0.99  $0.99/1ea  Fresh Vegetables   
3            Romaine Heart, 3 Pack   6.00   6.00  $2.00/1ea  Fresh Vegetables   
4                  Red Peppers       2.38   2.38  $9.90/1kg  Fresh Vegetables   

                5                                                  6  7      8  
0  20070132001_EA  https://www.realcanadiansuperstore.ca/english-...  1     ea  
1  20107500001_EA  https://www.realcanadiansuperstore.ca/green-on...  1  bunch  
2  20131170001_EA  https://www.realcanadiansuperstore.ca/bi-colou...  1     ea  
3  20067389001_EA  https://www.realcanadiansuperstore.ca/romaine-...  3   pack  
4  20007535001_KG  https://www.realcanadiansuperstore.ca/red-pepp...  1     ea

In [185]:
# add in header
df4_ec.columns = ['Name', 'Current Price', 'Previous Price', 'Price Per Each', 'Category', 'Product ID', 'Product URL', 'Amount', 'Units']

df4_ec.head()

Name Current Price Previous Price  \
0                 English Cucumber          1.99           1.99   
1                      Green Onion          1.29           1.29   
2  Bi-Colour Corn, Corn on the Cob          0.99           0.99   
3            Romaine Heart, 3 Pack          6.00           6.00   
4                  Red Peppers              2.38           2.38   

  Price Per Each          Category      Product ID  \
0      $1.99/1ea  Fresh Vegetables  20070132001_EA   
1      $1.29/1ea  Fresh Vegetables  20107500001_EA   
2      $0.99/1ea  Fresh Vegetables  20131170001_EA   
3      $2.00/1ea  Fresh Vegetables  20067389001_EA   
4      $9.90/1kg  Fresh Vegetables  20007535001_KG   

                                         Product URL Amount  Units  
0  https://www.realcanadiansuperstore.ca/english-...      1     ea  
1  https://www.realcanadiansuperstore.ca/green-on...      1  bunch  
2  https://www.realcanadiansuperstore.ca/bi-colou...      1     ea  
3  https://www.realcanadiansuperstore.ca/romaine-...      3   pack  
4  https://www.realcanadiansuperstore.ca/red-pepp...      1     ea

**Extra Credit (3 pts)**

1. Import a “dirty” dataset from the internet into your environment. (Missing values, improper coding of columns, etc.)
2. Clean or “tidy” the data when loading into python using pandas.

Looking at the datasets we have, `df2` has a column for salary which should ideally be an integer datatype and there is also a column of just missing values. Looking at the `csv` for this dataframe, the empty column is due to an extra comma after each salary entry. Let's change the format of the salary column to be more usable by converting to integer and remove the extra column. The state column is also improperly named as "Sate."

In [186]:
# clean df2
df2_cleaned = df2.rename(columns={"Sate ":"State"})

# change '2018 Mean Annual Wage' column to integer dtype
df2_cleaned['2018 Mean Annual Wage'] = df2_cleaned['2018 Mean Annual Wage'].apply(lambda x: int(x.replace(',', '')))

# remove empty column
df2_cleaned = df2_cleaned.drop('Unnamed: 2', axis=1)

Let's take a look now at the cleaned dataframe using `.head()`:

In [187]:
df2_cleaned.head()

State  2018 Mean Annual Wage
0    Alabama                   90620
1      Alaska                  99180
2     Arizona                  88800
3    Arkansas                  81430
4  California                  97110

We have removed the extra column and fixed the naming of the `State` column. 

Let's take a look at the datatypes of this cleaned dataframe:

In [188]:
df2_cleaned.dtypes

State                    object
2018 Mean Annual Wage     int64
dtype: object

The datatype for `2018 Mean Annual Wage` is now an integer. 

We can also now see descriptive statistics for this column:

In [189]:
df2_cleaned.describe()

2018 Mean Annual Wage
count              51.000000
mean            87563.333333
std              6117.818456
min             75010.000000
25%             83800.000000
50%             86690.000000
75%             90655.000000
max            107920.000000